### ISSUE TITLE
imap4 fetchSimplifiedBody fails for some emails

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4631]
|<img alt="robertimo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| robertimo reported|
|-|-|
|Trac ID|trac#4631|
|Type|defect|
|Created|2010-08-05 23:23:50Z|
|Branch|https://github.com/twisted/twisted/tree/bodystructure-4631-3|

Messages like the one in http://twistedmatrix.com/trac/ticket/4631#comment:5 confuse parsing of the BODY response. This is because it tries to do the parsing exclusively by looking at the response, without knowing what information was requested. In doing so, it fails to properly distinguish between the case where some header fields were requested (which it detects by looking to see if a particular part of the response is a list) and the case where the message was multipart/* containing multipart/* (which also results in that particular part of the response being a list).

One solution is to look at the result more carefully to differentiate between these two cases.

Attachments:

* [gmail.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/835/8351a2027976ef549f0fc2416068f1b883b3824e/e40fb7c6c3f739de25508d7441eb72e766b0cfb3.py) (1166 bytes) - added by robertimo on 2010-08-05 23:24:31Z - Script to generate the error
* [gmail-output](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/835/8351a2027976ef549f0fc2416068f1b883b3824e/3d607df18950141669de9a13cfdebe34c176cdb9) (1734 bytes) - added by robertimo on 2010-08-05 23:25:40Z - Output generated locally
* [fulltext](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/835/8351a2027976ef549f0fc2416068f1b883b3824e/d6b6dac26e160e7820cdb0647f5f38d04718bb48) (44714 bytes) - added by robertimo on 2010-08-06 02:11:03Z - The full text of an email that causes this problem.
* [4631-test.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/835/8351a2027976ef549f0fc2416068f1b883b3824e/bb73d3a1ff1af1ccc37dbd2f343ce69c7007e34e.patch) (1801 bytes) - added by atsuoi on 2010-09-07 04:12:45Z - patch against bodystructure-4631
* [4631-IMAP4FetchSimpliedBody-20120809.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/835/8351a2027976ef549f0fc2416068f1b883b3824e/e7a781b0677bfae584e1cfe9a87f0c73910e8dcb.patch) (6124733 bytes) - added by argonemyth on 2012-08-09 13:23:33Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4631 4631
type__defect defect
reporter__robertimo robertimo
priority__normal normal
milestone__ 
branch__branches_bodystructure_4631_3 branches/bodystructure-4631-3
branch_author__argonemyth__exarkun argonemyth, exarkun
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__imap4_email imap4 email
time__1281050630000000 1281050630000000
changetime__1359153044000000 1359153044000000
version__None None
owner__exarkun exarkun
cc__itamarst cc__atsuoi cc__argonemyth cc__rwall
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to robertimo|
|-|-|

Can you attach the actual email that triggers this?  Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

Also, what version of Twisted are you using? Version 10.1 fixed some bugs in the IMAP code, so try that if you're using an older version.

**Comment by twisted-trac** [comment]
|<img alt="robertimo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|robertimo commented|
|-|-|

I'm using Twisted 10.1.

**Comment by twisted-trac** [comment]
|<img alt="robertimo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|robertimo set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="atsuoi's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|atsuoi commented|
|-|-|

I could reproduce this problem with Twisted 10.1.

The mail trigered this problem has following structure.

```
MIME-Version: 1.0
Content-Type: multipart/mixed; boundary="----msg_border"
Date: Thu, 5 Aug 2010 00:49:45 -0700
Message-Id: <4C5A6D19.00000943@example.com>

This is a multi-part message in MIME format.

------msg_border
Date: Thu, 5 Aug 2010 00:49:45 -0700
Content-Type: multipart/alternative; boundary="----alt_border_1"

------alt_border_1
Content-Type: text/plain;
	charset="UTF-8"
Content-Transfer-Encoding: quoted-printable

TEXT PART OF MESSAGE.

------alt_border_1
Content-Type: text/html;
	charset="UTF-8"
Content-Transfer-Encoding: quoted-printable

<html><body></body></html>

------alt_border_1--

------msg_border--
```

For this mail, Gmail returns

    * 10 FETCH (UID 10 BODY ((("TEXT" "PLAIN" ("CHARSET" "UTF-8") NIL NIL "QUOTED-PRINTABLE" 10609 289) ("TEXT" "HTML" ("CHARSET" "UTF-8") NIL NIL "QUOTED-PRINTABLE" 31505 519) "ALTERNATIVE") "MIXED"))


t.m.imap4 parses this response as

```
[
    'UID', '10', 
    'BODY',
    [
         [
             ['TEXT', 'PLAIN', ['CHARSET', 'UTF-8'], None, None, 'QUOTED-PRINTABLE', '10609', '289'], 
             ['TEXT', 'HTML', ['CHARSET', 'UTF-8'], None, None, 'QUOTED-PRINTABLE', '31505', '519'], 
             'ALTERNATIVE'
         ],
         'MIXED'
    ]
]
```

Since the value of 'BODY' is a list of two items, IMAP4Client._parseFetchPairs() misinterpret the value as part of HEADER.FIELDS value or such.

I don't know proper way to fix problem, but IMAP4Client._parseFetchPairs() need stricter check.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [29974]) Branching to 'bodystructure-4631'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to robertimo|
|-|-|

I found and fixed some problems with multipart bodystructures.  However, I did not find a problem with `IMAP4Client.fetchSimplifiedBody`.

You can see the unit test I added here:

  http://twistedmatrix.com/trac/browser/branches/bodystructure-4631/twisted/mail/test/test_imap.py#L3650

If you can explain how the case you encountered differs from the one set up in that test (or even write a similar test that triggers the buggy behavior), that would be great.

**Comment by twisted-trac** [comment]
|<img alt="atsuoi's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|atsuoi commented|
|-|-|

I attached a patch against bodystructure-4631 branch to modify test to cause similar case. With this patch, _parseFetchPairs() raises ValueError(), which was silentry ignored.

**Comment by twisted-trac** [comment]
|<img alt="atsuoi's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|atsuoi removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @argonemyth|
|-|-|

There have been some changes to the IMAP4 _server_ BODYSTRUCTURE code.  I don't know if that's relevant to this ticket at all, but it probably makes sense to try merging the branch forward before doing anything else.

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)<a name="note_12"></a>|@argonemyth commented|
|-|-|

atsuoi is right about the problem in `_parseFetchPairs()`. The method was treating the following list:


```
[[['text', 'plain', None, None, None, None, '5', '1'], ['text', 'html', None, None, None, None, '6', '1'], 'alternative'], 'mixed']
```


as `["HEADER.FIELDS", ["SUBJECT"]]` because the length of the above list is less than 3.

I fix the problem by checking if it's a multipart message first:

```
# Check if value contains multpart message.
multipart = False
if value and isinstance(value[0], list):
    multipart = True 
...
if key in ("BODY", "BODY.PEEK") and isinstance(value, list) and len(value) < 3 and not multipart:
....
```

I also added the test supplied by atsuoi in test_imap.py, which now pass successfully.


About Merging Forward:

I did following to merge the Trunk to the branch, the resulted patch is nearly 6MB!

```
$ svn co svn://svn.twistedmatrix.com/svn/Twisted/branches/bodystructure-4631 Twisted
$ svn merge -r 29977:35091 svn://svn.twistedmatrix.com/svn/Twisted/trunk .
```

Two files have conflicts as you might guessed: `imap4.py` and `test_imap.py`. I merely accepted the trunk version and added my fix and the test after.

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)|@argonemyth removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

> I did following to merge the Trunk to the branch, the resulted patch is nearly 6MB!

This isn't something you should ever do.  You can find some documentation about Combinator, including merging forward, here: http://web.archive.org/web/20100107235656/http://divmod.org/trac/wiki/CombinatorTutorial

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)<a name="note_15"></a>|@argonemyth commented|
|-|-|

(In [35113]) Branching to 'bodystructure-4631-2'

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)<a name="note_16"></a>|@argonemyth commented|
|-|-|

Please review the branch bodystructure-4631-2! See comment:12 for more comments!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph commented|
|-|-|

argonemyth, I've forced some builds on the branch, [and you can see the results here](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/bodystructure-4631-2).  For information on how you can do this yourself in the future - it's generally a good idea to do it to get your branch reviewed faster - [see the committer checklist](https://twistedmatrix.com/trac/wiki/CommitterCheckList).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks.

  1. Considering the complexity of this code and of IMAP4 and the relative obscurity of the case in question, I think the test method docstring merits more detail than it currently has.
  1. Likewise, the comment near the changed code in the implementation is revealed as incomplete by this change and should probably be expanded.

It's too bad the `_parseFetchPairs` implementation isn't a bit more structured or rigorous, but Twisted itself is lacking the tools to support a better approach than is currently taken.  Perhaps one day someone will revisit this code and do a better job of it.

Since the two points above are just documentation issues, I'll address them and then merge the branch.  Thanks for your work on this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

(In [36730]) Expand the test method docstring, restructure to the implementation to be somewhat easier to follow (smaller expressions and more of them, with comments)

refs [#4631](https://github.com/twisted/twisted/issues/4631)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun removed owner|
|-|-|

Okay, I ended up doing a bit more than expanding the comment in the implementation.  I couldn't figure out how to explain the 5 part compound conditional guarding the extra `BODY` key/value processing code, so I refactored that code to make it a bit less dense and sprinkled some comments throughout it.  This merits another look by someone else, I think.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_22"></a>|@wallrj set owner to @exarkun<br>@wallrj set status to `new`|
|-|-|

# Code Review

I think this branch looks mostly ok.

Here are a few comments:

 1. Missing news file.
 2. The ticket summary doesn't make it clear that the problem is
    specifically to do with nested multipart messages (I think). Maybe
    write a clearer ticket summary - to match the message in the news
    file.
 3. Tests
    1. twisted.mail.tests all pass on my Fedora17 64 laptop.
    2. [http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/bodystructure-4631-2 Build Results]
       1. Looks like there are lots failures - are they expected?
 4. "_parseFetchPairs"
    1. There's a long TODO item mixed in with the @param documentation.
    2. It sounds similar to what exarkun said in
       ticket:4631#comment:18.
       ```
            ...(that is, just the part that comes after C{"FETCH"}).  The form
            of this input (and therefore the output of this method) is very
            disagreable.  A valuable improvement would be to enumerate the...
       ```
    3. So why not move that comment somewhere more appropriate and
       find or create a related trac ticket and add a link to the
       source code.
    4. The new commented BODY checking rules are much clearer, but...
    5. Comment says "BODY section has at most two elements" so why not
       "> 2" instead of ">= 3"?
 5. "_getContentType"
    1. why are you now lowercasing the mimetype?
    2. Is it related to this bug?
    3. If so, should there be a test?
    4. Looks like the output of _getContentType is lowercased again
       when its called in _getMessageStructure
 6. "test_fetchSimplifiedBodyMultipart"
    1. I found this test quite difficult to understand, especially
       when viewing it as a diff , but having looked at the rest of
       the file, I can see that it follows the style of other tests.
    2. I ran the new test against the old code and expected to see
       "IllegalServerResponse" as in the original bug report. I got
       this instead.

       ```
       Traceback (most recent call last):
         File "/home/richard/projects/Twisted/branches/bodystructure-4631-2/twisted/mail/test/test_imap.py", line 3730, in <lambda>
           d.addCallback(lambda x : self.assertEqual(self.result, self.expected))
         File "/home/richard/projects/Twisted/branches/bodystructure-4631-2/twisted/trial/_synctest.py", line 356, in assertEqual
           % (msg, pformat(first), pformat(second)))
       twisted.trial.unittest.FailTest: not equal:
       a = None
       b = {0: {'BODY': [[['text', 'plain', None, None, None, None, '5', '1'],
                      ['text', 'html', None, None, None, None, '6', '1'],
                      'alternative'],
                     'mixed']}}

       ```

       Is that expected?
    3. Should there be a separate tests for simple and nested
       multipart messages?

That's all.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun commented|
|-|-|

Messages like the one in http://twistedmatrix.com/trac/ticket/4631#comment:5 confuse parsing of the `BODY` response.  This is because it tries to do the parsing exclusively by looking at the response, without knowing what information was requested.  In doing so, it fails to properly distinguish between the case where some header fields were requested (which it detects by looking to see if a particular part of the response is a `list`) and the case where the message was multipart/* containing multipart/* (which also results in that particular part of the response being a `list`).

One solution is to look at the result more carefully to differentiate between these two cases.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun commented|
|-|-|

(In [36927]) Branching to 'bodystructure-4631-3'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun removed owner|
|-|-|

> It sounds similar to what exarkun said in ticket:4631#comment:18.

Is that the link you meant?  I don't see similar language in that comment.  Also, I agree that doesn't seem like the best place for that text, but this branch didn't add it.  Perhaps I should just file a new ticket for addressing it somehow, but leave the docstring alone for this ticket.

> _getContentType

The lowercasing looks unrelated.  I'm taking it out of the branch.

> I found this test quite difficult to understand, especially when viewing it as a diff , but having looked at the rest of the file, I can see that it follows the style of other tests.

Yes, not the best written test suite.  Sorry about that. :(

> I ran the new test against the old code and expected to see "IllegalServerResponse" as in the original bug report. I got this instead.

Seems like a bug in the helper code used to implement this test.  `self.connected` ends up with a `Failure` response wrapping that `IllegalServerResponse`.  But it has no errback to log the error, and since it is an attribute of the `TestCase` it does not get garbage collected in a timely manner.  I changed the code to explicitly log it, so you should see that exception if you repeat your experiment.

> Should there be a separate tests for simple and nested multipart messages?

Arguably.  A challenge here is that there is almost infinite complexity in MIME... Constructing a comprehensive collection of data to properly test this function is probably a big task, and not one I'm particularly keen to undertake at the moment.

All other points should be addressed.  Thanks for the review!

[Latest build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/bodystructure-4631-3)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_27"></a>|@wallrj set owner to @exarkun|
|-|-|

Replying to [exarkun](#note_26):
> > It sounds similar to what exarkun said in ticket:4631#comment:18.

> Is that the link you meant?  I don't see similar language in that
> comment.  Also, I agree that doesn't seem like the best place for
> that text, but this branch didn't add it.  Perhaps I should just
> file a new ticket for addressing it somehow, but leave the docstring
> alone for this ticket.

I meant These comments:

""" It's too bad the _parseFetchPairs implementation isn't a bit more
structured or rigorous """

""" A valuable improvement would be to enumerate the possible keys
(representing them as structured objects of some sort) rather than
using strings and tuples of tuples of strings and so forth.  """

Is there enough there to compose a new ticket? Maybe not. I'll leave
it up to you.

>> _getContentType
> The lowercasing looks unrelated.  I'm taking it out of the branch.

Ok.

>> I ran the new test against the old code and expected to see
>> "IllegalServerResponse" as in the original bug report. I got this
>> instead.
> Seems like a bug in the helper code used to implement this test.
> `self.connected` ends up with a `Failure` response wrapping that
> `IllegalServerResponse`.  But it has no errback to log the error,
> and since it is an attribute of the `TestCase` it does not get
> garbage collected in a timely manner.  I changed the code to
> explicitly log it, so you should see that exception if you repeat
> your experiment.

Yep, you're right. The original exception now shows up in the trial
output. That's much clearer.


>> Should there be a separate tests for simple and nested multipart messages?
> Arguably.  A challenge here is that there is almost infinite
> complexity in MIME... Constructing a comprehensive collection of
> data to properly test this function is probably a big task, and not
> one I'm particularly keen to undertake at the moment.

Understood.

> All other points should be addressed.  Thanks for the review!
> [Latest build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/bodystructure-4631-3)

Those results look much better than the -2 branch.

+1 from me.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_28"></a>|@exarkun commented|
|-|-|

> Is there enough there to compose a new ticket? Maybe not. I'll leave it up to you.

Maybe!  I gave it a try, [#6281](https://github.com/twisted/twisted/issues/6281).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun set status to `closed`|
|-|-|

(In [36933]) Merge bodystructure-4631-3

Author: atsuoi, argonemyth, exarkun
Reviewer: exarkun, rwall
Fixes: [#4631](https://github.com/twisted/twisted/issues/4631)

Fix `IMAP4Client` to be able to correctly parse the `BODY` response for a multipart/* message
which contains a multipart/* message.
